File: frontend/src/types/transaction.ts
export interface Transaction {
  id: number;
  type: TransactionType;
  holding_id: number | null;
  fd_id: number | null;
  amount: number;
  quantity: number | null;
  price: number | null;
  date: string;
  notes: string | null;
  source?: string;
  created_at: string;
}

export type TransactionType =
  | "deposit"
  | "withdrawal"
  | "buy"
  | "sell"
  | "dividend"
  | "interest"
  | "fd_open"
  | "fd_close"
  | "rebalance";

export interface TransactionCreate {
  type: TransactionType;
  amount: number;
  date: string;
  holding_id?: number;
  fd_id?: number;
  quantity?: number;
  price?: number;
  notes?: string;
}

export interface TransactionFilters {
  type?: string;
  holding_id?: string;
  fd_id?: string;
  date_from?: string;
  date_to?: string;
  limit?: string;
  offset?: string;
}

File: backend/app/schemas/transaction.py
import datetime as dt

from pydantic import BaseModel


class TransactionBase(BaseModel):
    type: str
    holding_id: int | None = None
    fd_id: int | None = None
    amount: float
    quantity: float | None = None
    price: float | None = None
    date: dt.date
    notes: str | None = None


class TransactionCreate(TransactionBase):
    source: str = "manual"


class TransactionUpdate(BaseModel):
    type: str | None = None
    holding_id: int | None = None
    fd_id: int | None = None
    amount: float | None = None
    quantity: float | None = None
    price: float | None = None
    date: dt.date | None = None
    notes: str | None = None


class TransactionResponse(TransactionBase):
    id: int
    realized_pnl: float | None = None
    source: str
    created_at: dt.datetime

    model_config = {"from_attributes": True}
